server/webapp/WEB-INF/rails.new/spec/controllers/api_v2/admin/environments_controller_spec.rb

```diff
@@ -0,0 +1,527 @@
+        expect(controller).to disallow_action(:put, :put, name: @environment_name).with(401, 'You are not authorized to perform this action.')
+      end
+
+      it 'should disallow normal users, with security enabled' do
+        login_as_user
+        expect(controller).to disallow_action(:put, :put, name: @environment_name).with(401, 'You are not authorized to perform this action.')
+      end
+    end
+
+    describe :route do
+      describe :with_header do
+        it 'should route to put action of environments controller for alphanumeric environment name' do
+          expect(:put => 'api/admin/environments/foo123').to route_to(action: 'put', controller: 'api_v2/admin/environments', name: 'foo123')
+        end
+
+        it 'should route to put action of environments controller for environment name with dots' do
+          expect(:put => 'api/admin/environments/foo.123').to route_to(action: 'put', controller: 'api_v2/admin/environments', name: 'foo.123')
+        end
+
+        it 'should route to put action of environments controller for environment name with hyphen' do
+          expect(:put => 'api/admin/environments/foo-123').to route_to(action: 'put', controller: 'api_v2/admin/environments', name: 'foo-123')
+        end
+
+        it 'should route to put action of environments controller for environment name with underscore' do
+          expect(:put => 'api/admin/environments/foo_123').to route_to(action: 'put', controller: 'api_v2/admin/environments', name: 'foo_123')
+        end
+
+        it 'should route to put action of environments controller for capitalized environment name' do
+          expect(:put => 'api/admin/environments/FOO').to route_to(action: 'put', controller: 'api_v2/admin/environments', name: 'FOO')
+        end
+      end
+      describe :without_header do
+        before :each do
+          teardown_header
+        end
+        it 'should not route to put action of environments controller without header' do
+          expect(:put => 'api/admin/environments/foo').to_not route_to(action: 'put', controller: 'api_v2/admin/environments')
+          expect(:put => 'api/admin/environments/foo').to route_to(controller: 'application', action: 'unresolved', url: 'api/admin/environments/foo')
+        end
+      end
+    end
+  end
+
+  describe :patch do
+
+    before(:each) do
+      @environment_name = 'foo-environment'
+      @environment_config = BasicEnvironmentConfig.new(CaseInsensitiveString.new(@environment_name))
+      @environment_config_service = double('environment-config-service')
+      controller.stub(:environment_config_service).and_return(@environment_config_service)
+      @environment_config_service.stub(:getEnvironmentConfig).with(@environment_name).and_return(@environment_config)
+    end
+
+    describe :for_admins do
+      it 'should allow patching environments' do
+        login_as_admin
+        result = HttpLocalizedOperationResult.new
+        pipelines_to_add = ['foo']
```

Comment:
You could have a before each block if this setup is the same for the 2 specs in for_admins block